### PR TITLE
feat: Automatic final model upload to S3 after training

### PR DESCRIPTION
**Pull request by efazal**
**What this PR does / why we need it**:

Automatically uploads final model artifacts to S3 after the user's training code completes. This ensures the final trained model is persisted to remote storage alongside periodic checkpoints.                                     
                                                                                                                                                                                                                                        
  - Add upload_final_model_to_cloud() as a post-training cleanup function that runs after user code finishes (via generated footer code)
  - Upload all remaining files/directories in output_dir to S3, preserving the exact local filesystem structure
  - Skip already-uploaded checkpoint-* dirs, internal cloud-upload-staging dir, and .cache
  - Only local rank 0 uploads to avoid race conditions in distributed training

Node 0:
<img width="707" height="247" alt="image" src="https://github.com/user-attachments/assets/cfe16e6a-151b-49df-9c5c-2b9f24a5ccb2" />

Node 1:
<img width="713" height="295" alt="image" src="https://github.com/user-attachments/assets/38d0358a-19d9-4687-a8bb-5a281579e43e" />



<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Final models are now uploaded to cloud storage automatically after training/save; uploads run as pre/post instrumentation and are gated to the primary process.

* **Public API**
  * Background upload lifecycle is now exposed and more robust, improving reliability of uploads, restarts and shutdown behavior.
  * Instrumentation now uses distinct header/footer hooks (two-part injection) and small API surface adjustments to control uploads.

* **Tests**
  * Expanded coverage for final-model upload, restart/resume logic, worker lifecycle, storage interactions and edge cases.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

### CODE REVIEW

File: kubeflow/trainer/rhai/transformers.py [hunk 1]
```diff
@@ -492,12 +492,15 @@ def _check_upload_error(self) -> None:
                     self._upload_error = None  # Clear after reading
                     raise error
 
-        def _start_upload_worker(self) -> None:
+        def start_upload_worker(self) -> None:
             """Start background upload worker."""
-            if self._upload_queue is None:
+            # Start worker if queue doesn't exist OR if the worker thread is not alive
+            upload_thread = getattr(self, "_upload_thread", None)
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: nit: ?
```suggestion
            upload_thread = self._upload_thread
```

File: kubeflow/trainer/rhai/transformers.py [hunk 2]
```diff
@@ -877,6 +880,63 @@ def apply_checkpointing():
...
+
+                Only uploads when _internal_call=False (user-initiated saves like final model).
+                Periodic checkpoints (_internal_call=True) are already handled by JIT checkpoint infrastructure.
+                """
+                # Call original save_model (handles all distributed training logic)
+                result = _original_save_model(
+                    self,
+                    output_dir=output_dir,
+                    _internal_call=_internal_call,
+                )
```
[the remark this answers is not in the document]
Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py]: @efazal fyi - Transformers have changed this API in the past.

File: kubeflow/trainer/rhai/transformers.py [hunk 3]
```diff
@@ -877,6 +880,63 @@ def apply_checkpointing():
...
+
+                        # Queue upload using existing infrastructure
+                        total_size = _jit_checkpoint_callback.calculate_local_dir_size(output_dir)
+                        _jit_checkpoint_callback.start_upload_worker()
+
+                        node = os.environ.get("NODE_RANK") or os.environ.get("GROUP_RANK") or "0"
+                        marker_name = f"{CHECKPOINT_INCOMPLETE_MARKER}.node-{node}-rank-{self.args.local_process_index}"
+
+                        task = (output_dir, checkpoint_name, total_size, marker_name)
+                        _jit_checkpoint_callback.upload_queue.put(task)
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: do we need this running in a background thread? We're adding the task then immediately waiting for the task to complete.

File: kubeflow/trainer/rhai/transformers.py [hunk 4]
```diff
@@ -877,6 +880,63 @@ def apply_checkpointing():
...
+                    self,
+                    output_dir=output_dir,
+                    _internal_call=_internal_call,
+                )
+
+                # Upload after save ONLY for user-initiated saves (final model, not periodic checkpoints)
+                if (
+                    not _internal_call
+                    and _jit_checkpoint_callback.remote_fs
+                    and output_dir
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: won't this mean no model is uploaded to S3 if the user doesn't specify output_dir and it defaults to None? Doesn't transformers automatically write the model to the configured `args.output_dir` if no arg is given?

File: kubeflow/trainer/rhai/transformers.py [hunk 5]
```diff
@@ -877,6 +880,63 @@ def apply_checkpointing():
...
+                    and os.path.exists(output_dir)
+                ):
+                    # Barrier to ensure all ranks finished saving
+                    _jit_checkpoint_callback.wait_for_all_ranks("final_save")
+
+                    # Upload from local rank 0 only (works for all strategies - DDP/FSDP/DeepSpeed)
+                    if self.args.local_process_index == 0:
+                        # Determine checkpoint name from output_dir
+                        checkpoint_name = os.path.basename(output_dir)
```
Comment (robert-bell) [inline kubeflow/trainer/rhai/transformers.py]: Won't this cause weird behaviour if output_dir arg is not a subpath of the "mounted" s3 bucket?

**Review by robert-bell — CHANGES_REQUESTED** [review]
As I mentioned on our call, monkey patching is much more brittle. I do think we should go with the post-train instrumentation that copies any files remaining in the empty dir up to s3.

File: kubeflow/trainer/rhai/transformers.py [hunk 6]
```diff
@@ -877,6 +880,63 @@ def apply_checkpointing():
...
+
+                        task = (output_dir, checkpoint_name, total_size, marker_name)
+                        _jit_checkpoint_callback.upload_queue.put(task)
+                        _log(
+                            f"[Kubeflow] Queued final model for upload: {checkpoint_name}",
+                            args=self.args,
+                        )
+
+                        # Wait for final model upload to complete
+                        _jit_checkpoint_callback.shutdown_upload_worker()
```
Comment (hrathina) [inline kubeflow/trainer/rhai/transformers.py]: Do we need _jit_checkpoint_callback.shutdown_upload_worker() here, since we’re already calling shutdown_upload_worker() in on_train_end()?

File: kubeflow/trainer/rhai/transformers.py [hunk 7]
```diff
@@ -1054,11 +1059,80 @@ def _patched_train(resume_from_checkpoint=None, **train_kwargs):
...
+        Works for all distributed training strategies (DDP, FSDP, DeepSpeed)
+        by uploading from local rank 0 on each node. Files already uploaded
+        as periodic checkpoints are skipped.
+        """
+        if not checkpoint_config.get("cloud_remote_storage_uri"):
+            return
+
+        local_rank = int(os.environ.get("LOCAL_RANK", "0"))
+        if local_rank != 0:
+            return
```
[the remark this answers is not in the document]
Comment (robert-bell) [reply kubeflow/trainer/rhai/transformers.py:1099]: @efazal does this need fixing?

**Review by robert-bell — COMMENTED** [review]
lgtm

@hrathina please can you review the tests?

**Review by hrathina — APPROVED** [review]
